server/app/src/main/java/com/google/heartrate/wearos/app/gatt/heartrate/characteristics/HeartRateMeasurementCharacteristic.java

```diff
@@ -0,0 +1,269 @@
+     * @param expendedEnergy       (kiloJoules) expended energy value in UInt16 format
+     * @throws GattException in case of wrong arguments format or offset for heart rate characteristic values
+     */
+    public void setHeartRateCharacteristicValue(int heartRateMeasurement, int expendedEnergy) throws GattException {
+        Log.i(TAG, String.format("Set the heart measurements value %d and expended energy value %d",
+                heartRateMeasurement, expendedEnergy));
+
+        FormatUtils.assertIsUInt16(heartRateMeasurement);
+        FormatUtils.assertIsUInt16(expendedEnergy);
+
+        if (FormatUtils.isInUInt8Range(heartRateMeasurement)) {
+
+            /* Flags(UInt8) + HR(UInt8) + EE(UInt16) */
+            setValue(new byte[4]);
+
+            /* EE present and HR UInt8 format => Flags = 0b00001000 */
+            setIntValue((byte) (HEART_RATE_MEASUREMENT_UINT8_FLAG | EXPENDED_ENERGY_FLAG),
+                    BluetoothGattCharacteristic.FORMAT_UINT8,
+                    FLAGS_OFFSET);
+
+            setIntValue(heartRateMeasurement,
+                    BluetoothGattCharacteristic.FORMAT_UINT8,
+                    HEART_RATE_MEASUREMENT_OFFSET);
+
+            /* HR is in UInt8 format => EE offset is base */
+            setIntValue(expendedEnergy,
+                    BluetoothGattCharacteristic.FORMAT_UINT16,
+                    EXPENDED_ENERGY_BASE_OFFSET);
+        } else {
+
+            /* Flags(UInt8) + HR(UInt16) + EE(UInt16) */
+            setValue(new byte[5]);
+
+            /* EE present and HR UInt16 format => Flags = 0b00001001 */
+            setIntValue((byte) (HEART_RATE_MEASUREMENT_UINT16_FLAG | EXPENDED_ENERGY_FLAG),
+                    BluetoothGattCharacteristic.FORMAT_UINT8,
+                    FLAGS_OFFSET);
+
+            setIntValue(heartRateMeasurement,
+                    BluetoothGattCharacteristic.FORMAT_UINT16,
+                    HEART_RATE_MEASUREMENT_OFFSET);
+
+            /* HR is in UInt16 format => EE offset is shifted */
+            setIntValue(expendedEnergy,
+                    BluetoothGattCharacteristic.FORMAT_UINT16,
+                    EXPENDED_ENERGY_SHIFTED_OFFSET);
+        }
+    }
+
+    /**
+     * Set given Heart Rate Measurement value to the characteristic.
+     * <p>
+     * Flags 8bit format 0b00000000 from right to left:
+     * <br>Heart Rate Format (0) -> UInt8 Format, (1) -> UInt16 Format
+     * <br>Sensor Contact Status (00) -> Not Supported
+     * <br>Energy Expended (0) -> Field not Present
+     * <br>RR-Interval (0) -> Field not presented
+     * <br>Unused (000)
+     *
+     * @param heartRateMeasurement (bpm) heart rate measurement value in UInt8/UInt16 format
+     * @throws GattException in case of wrong arguments format or offset for heart rate characteristic values
+     */
+    public void setHeartRateCharacteristicValue(int heartRateMeasurement) throws GattException {
```

Comment:
There is code duplication between this method and the previous one. Is there a way to call `public void setHeartRateCharacteristicValue(int heartRateMeasurement, int expendedEnergy)` from this one?

Reply:
I don't think it is a good idea, this two separate methods look much clearer in terms of protocol.

Reply:
I agree with Idriss. There is a lot of code duplication. How about a private method that takes heartRateMeasurement int and an Optional<Integer> for the expended energy, and call that method from both public methods?

Reply:
Ok, I'll do like this, but if you look at these two methods more detailed, you'll see, that they operate quite different values, which are very dependent to each other.

Reply:
Thank you. Marking this as resolved.